velox/common/encode/Base32.cpp

```diff
@@ -0,0 +1,490 @@
+ */
+#include "velox/common/encode/Base32.h"
+
+#include <folly/Expected.h>
+#include <folly/container/Foreach.h>
+#include <folly/io/Cursor.h>
+#include <cctype>
+#include <cstdint>
+
+#include "velox/common/base/Exceptions.h"
+
+namespace facebook::velox::encoding {
+
+// Character set for Base32 encoding (RFC 4648)
+constexpr const Base32::Charset kBase32Charset = {
+    'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K',
+    'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V',
+    'W', 'X', 'Y', 'Z', '2', '3', '4', '5', '6', '7'};
+
+// Reverse lookup table for decoding. 255 means invalid character.
+constexpr const Base32::ReverseIndex kBase32ReverseIndexTable = {
+    255, 255, 255, 255, 255, 255, 255, 255,
+    255, 255, 255, 255, 255, 255, 255, 255, // 0-15
+    255, 255, 255, 255, 255, 255, 255, 255,
+    255, 255, 255, 255, 255, 255, 255, 255, // 16-31
+    255, 255, 255, 255, 255, 255, 255, 255,
+    255, 255, 255, 255, 255, 255, 255, 255, // 32-47
+    255, 255, 26,  27,  28,  29,  30,  31,
+    255, 255, 255, 255, 255, 255, 255, 255, // 48-63 ('2'-'7')
+    255, 0,   1,   2,   3,   4,   5,   6,
+    7,   8,   9,   10,  11,  12,  13,  14, // 64-79 ('A'-'O')
+    15,  16,  17,  18,  19,  20,  21,  22,
+    23,  24,  25,  255, 255, 255, 255, 255, // 80-95 ('P'-'Z')
+    255, 0,   1,   2,   3,   4,   5,   6,
+    7,   8,   9,   10,  11,  12,  13,  14, // 96-111 ('a'-'o')
+    15,  16,  17,  18,  19,  20,  21,  22,
+    23,  24,  25,  255, 255, 255, 255, 255, // 112-127 ('p'-'z')
+    255, 255, 255, 255, 255, 255, 255, 255,
+    255, 255, 255, 255, 255, 255, 255, 255, // 128-143
+    255, 255, 255, 255, 255, 255, 255, 255,
+    255, 255, 255, 255, 255, 255, 255, 255, // 144-159
+    255, 255, 255, 255, 255, 255, 255, 255,
+    255, 255, 255, 255, 255, 255, 255, 255, // 160-175
+    255, 255, 255, 255, 255, 255, 255, 255,
+    255, 255, 255, 255, 255, 255, 255, 255, // 176-191
+    255, 255, 255, 255, 255, 255, 255, 255,
+    255, 255, 255, 255, 255, 255, 255, 255, // 192-207
+    255, 255, 255, 255, 255, 255, 255, 255,
+    255, 255, 255, 255, 255, 255, 255, 255, // 208-223
+    255, 255, 255, 255, 255, 255, 255, 255,
+    255, 255, 255, 255, 255, 255, 255, 255, // 224-239
+    255, 255, 255, 255, 255, 255, 255, 255,
+    255, 255, 255, 255, 255, 255, 255, 255 // 240-255
+};
+
+// static
+folly::Expected<uint8_t, Status> Base32::base32ReverseLookup(
+    char encodedChar,
+    const ReverseIndex& reverseIndex) {
+  auto index = reverseIndex[static_cast<uint8_t>(encodedChar)];
+  if (index >= 0x20) {
+    return folly::makeUnexpected(Status::UserError(
+        "decode() - invalid input string: invalid character '{}'",
+        encodedChar));
+  }
+  return index;
+}
+
+// static
+size_t Base32::calculateEncodedSize(size_t inputSize, bool withPadding) {
+  if (inputSize == 0) {
+    return 0;
+  }
+
+  // Base32 encodes 5 bytes into 8 characters
+  size_t encodedSize = ((inputSize + Base32::kBinaryBlockByteSize - 1) /
```

Comment:
nit: Is there risk of overflow underflow due to division and multiplication ? Should you use checkedMultiply / divide.